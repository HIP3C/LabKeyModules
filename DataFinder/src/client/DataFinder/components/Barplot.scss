.datafinder-barplot {
  overflow-y: auto;
  // border-top: 1px solid #dddddd;
  // border-bottom: 1px solid #dddddd
}

.df-barplot-title {
  text-align: center;
  margin-bottom: 0;
}
