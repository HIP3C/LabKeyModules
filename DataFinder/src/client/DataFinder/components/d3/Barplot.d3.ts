import * as d3 from 'd3'
import { PlotDatum } from '../../../typings/CubeData';

// ================================================================== //
/* 
This is a barplot component which takes the following arguments in 
the props: 
  * data
  * name (id for the div to stick it in)
  * width
  * height

*/
// Types

interface D3BarplotConfig {
    width: number;
    height: number;
    labels: string[];
    countMetric: string;
    barColor: string;
}


interface D3Barplot {
    create: (id: string, data: PlotDatum[], config: D3BarplotConfig) => void;
    update: (id: string, data: PlotDatum[], config: D3BarplotConfig) => void;
}


const xaxisTitle = {height: 40}

const createBarplot = (id: string, data: PlotDatum[], config: D3BarplotConfig) => {

    // Create margins
    const barplotHeight = config.height - xaxisTitle.height - 10
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
    width = config.width - margin.left - margin.right,
    height = barplotHeight - margin.top - margin.bottom;

    // Add svg for bars and separate one for x-axis to allow
    const svg = d3
        .select("#barplot-container-" + id).select("svg")
        .attr("height", barplotHeight)
        .attr("width", config.width)
        .attr("id", "svg-barplot-" + id)

    const xAxisSvg = d3.select("#xaxis-" + id).select("svg")
        .attr("width", config.width)
        .attr("height", xaxisTitle.height)


    // Get long and short versions of y-axis labels
    const labels = config.labels.map(l => {
        let short = l
        const maxLength = Math.round( margin.left / 5.5 )
        if (l.length > maxLength) short = l.slice(0,  maxLength - 3) + "..."
        return ({ label: l, shortlabel: short })
    })
    const yaxisScale = d3
        .scaleBand()
        .domain(labels.map(l => l.label))
        .range([height, 0]);

    // Add container for bars
    const barplot = svg
        .append("g")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .attr("id", "bars-" + id)
        .attr("style", "overflow:auto")

    // Add y-axis short labels
    const yaxisLabels = svg.append("g")
        .attr("id", "yaxis-labels-short-" + id)
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
    yaxisLabels.selectAll("text")
        .data(labels)
        .enter()
        .append("text")
        .attr("x", -5)
        .attr("y", (d) => {
            return yaxisScale((d.label)) + yaxisScale.bandwidth() / 1.5;
        })
        .attr("text-anchor", "end")
        .attr("font-size", ".8em")
        .text(d => d.shortlabel)

    // Insert vertical grid lines (before bars)
    const xaxisGrid = svg.insert("g", ":first-child")
        .attr("id", "xaxis-grid-" + id)
        .attr(
            "transform",
            `translate(${margin.left}, ${(height + 5 + margin.top)})`
        )

    // x-axis label container (scale changes on update)
    const xaxisLabels = xAxisSvg.append("g")
        .attr("class", "x axis")
        .attr(
            "transform",
            `translate(${margin.left}, 0)`
        )

    // x-axis title
    xAxisSvg.append("g")
        .attr(
            "transform",
            `translate(${margin.left}, 0)`
        )
        .append("text")
        .attr("x", (width) / 2)
        .attr("y", 30)
        .attr("class", "x-axis-title")
        .text(config.countMetric == "studyCount" ? "Studies" : "Participants")
        .attr("text-anchor", "middle")
        .attr("font-size", "0.8em")

    // y-axis long labels container (for hover)
    const yaxisLongLabels = svg.append("g")
        .attr("id", "yaxis-labels-" + id)
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
}

const updateBarplot = (id: string, data: PlotDatum[], config: D3BarplotConfig) => {
    // Prelims -----------------
    const svg = d3.select("#barplot-container-" + id).select("svg")
    const xAxisSvg = d3.select("#xaxis-" + id).select("svg")
    const barplot = d3.select("#bars-" + id)
    const xaxisGrid = svg.select("#xaxis-grid-" + id)
    const xaxisLabels = xAxisSvg.selectAll(".x.axis")
    // yaxisLongLabels -- a group on top of the othe
    const yaxisLongLabels = svg.select("#yaxis-labels-" + id) 

    

    // Create margins
    const barplotHeight = config.height - xaxisTitle.height - 10
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
    width = config.width - margin.left - margin.right,
    height = barplotHeight - margin.top - margin.bottom;
    
    const yaxisScale = d3
        .scaleBand()
        .domain(config.labels.map(l => l))
        .range([height, 0]);
    

    // Make the updates ------------
    // Bind data
    // Get range of data for x-axis
    const dataRange = [0, 10];
    const countMetric = config.countMetric
    data.forEach((v) => (v[countMetric] > dataRange[1]) && (dataRange[1] = v[countMetric]))
    const xaxisScale = d3
        .scaleLinear()
        .domain(dataRange).nice()
        .range([0, width]);
    const updateXaxis: any = d3.axisBottom(xaxisScale).ticks(3).tickSize(0)
    const updateXaxisGrid: any = d3.axisBottom(xaxisScale).ticks(3).tickSize(-(height + 5)).tickFormat(() => "")

    // Bind the data
    const boxes = barplot.selectAll("rect").data(data);

    const yaxisHoverLabels = yaxisLongLabels
        .selectAll(".yaxis-long-label-container")
        .data(data)

    // ---- enter ----
    // bars on the barplot
    boxes
        .enter()
        .append("rect")
        .attr("class", "rect")
        .attr("x", xaxisScale(0))
        .attr("width", function (d: PlotDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: PlotDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)
        .style("fill", config.barColor)
    
    // Labels for hover
    const labelContainers = yaxisHoverLabels
        .enter()
        .append("g")
        .attr("class", "yaxis-long-label-container") 

    labelContainers
        .append("text")
        .attr("class", "label-number")
        .attr("x", config.width - margin.left)
        .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
        .attr("text-anchor", "end")
        .text(d => d[countMetric] == null ? 0 : d[countMetric])
        .attr("font-size", ".8em")
        .attr("fill", "transparent")
        
    // Rect over the whole area around the bar to use for adding mouseover event
    labelContainers
        .append("rect")
        .attr("class", "big-rect-cover")
        .attr("x", - margin.left)
        .attr("y", (d: PlotDatum) => {return(yaxisScale(d.member))})
        .attr("width", config.width)
        .attr("height", yaxisScale.bandwidth())
        .attr("fill", "transparent")
        .attr("stroke", "transparent")
    // rect over the short label. Will become opaque to cover this label
    // and replace with the long version on mouseover. 
    labelContainers
        .append("rect")
        .attr("class", "little-rect-cover")
        .attr("x", d => d == null ? undefined : -margin.left)
        .attr("y", (d: PlotDatum) => yaxisScale(d.member))
        .attr("width", 100)
        .attr("height", yaxisScale.bandwidth())
        .attr("fill", "transparent")
        .attr("stroke", "transparent")
    // y-axis labels
    labelContainers
        .append("text")
        .attr("class", "label-text")
        .attr("x", (d: PlotDatum) => d.member.length < 18 ? -5 : - margin.left)
        .attr("y", (d: PlotDatum) => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
        .attr("text-anchor", (d: PlotDatum) => d.member.length < 18 ? "end" : "start")
        .attr("font-size", ".8em")
        .attr("fill", "transparent")
      .selectAll("tspan")
        .data((d: PlotDatum) => {
            let l = d.member
            const maxLength = Math.round( (config.
                width - margin.right - 10) / 5.5 )
            if (l.length > maxLength) {
                let splitIndex = l.indexOf(" ")
                while (splitIndex < maxLength) {
                    const next = l.indexOf(" ", splitIndex + 1)
                    if (next > -1 && next < maxLength) { splitIndex = next } else { break }
                }
                let newlabel = [l.slice(0, splitIndex), l.slice(splitIndex + 1)]
                if (newlabel[1].length > maxLength) newlabel[1] = newlabel[1].slice(0, maxLength - 4) + "..."
                return (newlabel)
            } else { return ([l]) }
        })
        .enter()
        .append("tspan")
        .attr("x", (d, i) => d.length < 18 && i === 0 ? -5 : - margin.left)
        .attr("dy", (d, i, data) => i > 0 ? "10" : data.length > 1 ? "-5" : "0")
        .text(d => d)
    
    

    // ---- update ----
    boxes
        .transition()
        .duration(500)
        .attr("width", function (d: PlotDatum) {
            return xaxisScale(d[countMetric]);
        })

    yaxisHoverLabels
        .selectAll(".label-number")
        .data(d => [d])
        .text(d => d[countMetric] == null ? 0 : d[countMetric])

    xaxisLabels.transition()
        .duration(500)
        .call(updateXaxis)
    xaxisGrid.transition()
        .duration(500)
        .call(updateXaxisGrid)
        .selectAll("line").attr("stroke", "#ddd")
    d3.selectAll(".domain").remove() // Removes the line on the bottom of the plot

    // ---- exit ----
    boxes.exit().remove();

    // ----- Mouseover event -----
    yaxisLongLabels.selectAll("g")
        .on("mouseover", function (d) {
            d3.select(this)
                .selectAll(".label-text")
                .transition()
                .duration(100)
                .attr("fill", "black")
                .attr("font-weight", "bold")
            d3.select(this)
                .selectAll(".label-number")
                .transition()
                .duration(100)
                .attr("fill", "black")
                .attr("font-weight", "bold")
            d3.select(this)
                .selectAll(".little-rect-cover")
                .transition()
                .duration(100)
                .attr("fill", "white")
        })
        .on("mouseout", function (d) {
            d3.select(this)
                .selectAll(".label-text")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
                .attr("font-weight", "normal")
            d3.select(this)
                .selectAll(".label-number")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
                .attr("font-weight", "normal")
            d3.select(this)
                .selectAll(".little-rect-cover")
                .transition()
                .duration(100)
                .attr("fill", "transparent")
        })

}

export const D3Barplot: D3Barplot = {
    create: createBarplot,
    update: updateBarplot
}
