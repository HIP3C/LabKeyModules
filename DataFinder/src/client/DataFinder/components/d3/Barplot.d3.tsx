import * as d3 from 'd3'
import { CubeDatum } from '../../../typings/CubeData';
import { entries } from 'd3';

// ================================================================== //
/* 
This is a barplot component which takes the following arguments in 
the props: 
  * data
  * name (id for the div to stick it in)
  * width
  * height

*/
// Types
interface DrawBarplotProps {
    data: CubeDatum[];
    name: string;
    width: number;
    height: number;
    labels: string[];
    countMetric: string;
    barColor: string;
}


export function drawBarplot(props: DrawBarplotProps) {

    const data = props.data;
    const name = props.name;

    const dataRange = [0, 10];
    const countMetric = props.countMetric
    data.forEach((v) => (v[countMetric] > dataRange[1]) && (dataRange[1] = v[countMetric]))

    const totalHeight = Math.max(195, 15 * props.labels.length + 20)

    const labels = props.labels.map(l => {
        let short = l
        if (l.length > 18) short = l.slice(0, 14) + "..."
        return ({ label: l, shortlabel: short })
    })
    // debuggers
    // const newLabels = [];

    // data.map((e, i) => {
    //     newLabels.push(e.label);
    //     // if (e.value > dataRange[1]) dataRange[1] = e.value;
    // });

    const svg = d3
        .select("#barplot-container-" + name).select("svg")
        .attr("height", totalHeight)
        .attr("width", 220)
        .attr("id", "barplot-container-" + name)

    const xAxisSvg = d3.select("#xaxis-" + name).select("svg")
        .attr("width", props.width)
        .attr("height", 40)


    // Create margins
    const margin = { top: 0, right: 15, bottom: 0, left: 100 },
        width = props.width - margin.left - margin.right,
        height = totalHeight - margin.top - margin.bottom;

    // Set scales using options

    const xaxisScale = d3
        .scaleLinear()
        .domain(dataRange).nice()
        .range([0, width]);

    const yaxisScale = d3
        .scaleBand()
        .domain(labels.map(l => l.label))
        .range([height, 0]);

    // Create body and axes
    // svg.append("g")
    //     .call(d3.axisLeft(yaxisScale));

    // svg.append("g")
    //     .call(d3.axisBottom(xaxisScale));

    let barplot: d3.Selection<SVGElement, any, HTMLElement, any>;
    let xaxisGrid, xaxisLabels, yaxisLabels
    if (!svg.selectAll("g").empty()) {
        svg.select("#xaxis-labels-" + name).remove()
        xaxisGrid = svg.select("#xaxis-grid-" + name)
        xaxisLabels = xAxisSvg.selectAll(".x.axis")
        barplot = svg.select("#barplot" + name)
    } else {
        barplot = svg
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
            .attr("id", "barplot" + name)
            .attr("style", "overflow:auto")

        yaxisLabels = svg.append("g")
            .attr("id", "yaxis-labels-short-" + name)
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
        yaxisLabels.selectAll("text")
            .data(labels)
            .enter()
            .append("text")
            .attr("x", -5)
            .attr("y", (d) => {
                return yaxisScale((d.label)) + yaxisScale.bandwidth() / 1.5;
            })
            .attr("text-anchor", "end")
            .attr("font-size", ".8em")
            .text(d => d.shortlabel)

        const yaxisLongLabels = svg.append("g")
            .attr("id", "yaxis-labels-" + name)
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        const labelContainers = yaxisLongLabels
            .selectAll("g")
            .data(data)
            .enter()
            .append("g")

        labelContainers
            .selectAll(".big-rect-cover")
            .data(d => [d])
            .enter()
            .append("rect")
            .attr("class", "big-rect-cover")
            .attr("x", - margin.left)
            .attr("y", d => yaxisScale(d.member))
            .attr("width", props.width)
            .attr("height", yaxisScale.bandwidth())
            .attr("fill", "transparent")
            .attr("stroke", "transparent")
        labelContainers
            .selectAll(".little-rect-cover")
            .data(d => [d])
            .enter()
            .append("rect")
            .attr("class", "little-rect-cover")
            .attr("x", d => d == null ? undefined : -margin.left)
            .attr("y", d => yaxisScale(d.member))
            .attr("width", 100)
            .attr("height", yaxisScale.bandwidth())
            .attr("fill", "transparent")
            .attr("stroke", "transparent")
        labelContainers.selectAll(".label-text")
            .data(d => [d])
            .enter()
            .append("text")
            .attr("class", "label-text")
            .attr("x", d => d.member.length < 18 ? -5 : - margin.left)
            .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
            .attr("text-anchor", d => d.member.length < 18 ? "end" : "start")
            .attr("font-size", ".8em")
            .attr("fill", "transparent")
            .text(d => d.member)
        labelContainers.selectAll(".label-number")
            .data(d => [d])
            .enter()
            .append("text")
            .attr("class", "label-number")
            .attr("x", props.width - margin.left - 30)
            .attr("y", d => yaxisScale((d.member)) + yaxisScale.bandwidth() / 1.5)
            .attr("text-anchor", "end")
            .text(d => d[countMetric] == null ? 0 : d[countMetric])
            .attr("font-size", ".8em")
            .attr("fill", "transparent")

        yaxisLongLabels.selectAll("g")
            .on("mouseover", function (d) {
                d3.select(this)
                    .selectAll(".label-text")
                    .transition()
                    .duration(100)
                    .attr("fill", "black")
                    .attr("font-weight", "bold")
                d3.select(this)
                    .selectAll(".label-number")
                    .transition()
                    .duration(100)
                    .attr("fill", "black")
                    .attr("font-weight", "bold")
                d3.select(this)
                    .selectAll(".little-rect-cover")
                    .transition()
                    .duration(100)
                    .attr("fill", "white")
            })
            .on("mouseout", function (d) {
                d3.select(this)
                    .selectAll(".label-text")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
                    .attr("font-weight", "normal")
                d3.select(this)
                    .selectAll(".label-number")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
                    .attr("font-weight", "normal")
                d3.select(this)
                    .selectAll(".little-rect-cover")
                    .transition()
                    .duration(100)
                    .attr("fill", "transparent")
            })

        xaxisGrid = svg.insert("g", ":first-child")
            .attr("id", "xaxis-grid-" + name)
            .attr(
                "transform",
                `translate(${margin.left}, ${(height + 5 + margin.top)})`
            )

        xaxisLabels = xAxisSvg.append("g")
            .attr("class", "x axis")
            .attr(
                "transform",
                `translate(${margin.left}, 0)`
            )

        xAxisSvg.append("g")
            .attr(
                "transform",
                `translate(${margin.left}, 0)`
            )
            .append("text")
            .attr("x", (width) / 2 )
            .attr("y", 30)
            .text(props.countMetric == "studyCount" ? "Studies" : "Participants")
            .attr("text-anchor", "middle")
            .attr("font-size", "0.8em")
    }

    xaxisGrid.transition()
        .duration(500)
        .call(d3.axisBottom(xaxisScale).ticks(3).tickSize(-(height + 5)).tickFormat(() => ""))
        .selectAll("line").attr("stroke", "#ddd")
    xaxisLabels.transition()
        .duration(500)
        .call(d3.axisBottom(xaxisScale).ticks(3).tickSize(0))
    // svg.append("g")
    //     .attr("id", "xaxis-labels-" + name)
    //     .attr(
    //         "transform",
    //         `translate(${margin.left}, ${(height + margin.top)})`
    //     )
    //     .call(d3.axisBottom(xaxisScale).tickValues([Math.round((dataRange[1] + 5) / 10) * 10]).tickSize(-height))
    //     .selectAll("line").attr("stroke", "#bcbcbc").attr("stroke-width", "3")

    d3.selectAll(".domain").remove()




    // add data
    const boxes = barplot.selectAll("rect").data(data);
    boxes
        .enter()
        .append("rect")
        .attr("class", "rect")
        .attr("x", xaxisScale(0))
        .attr("width", function (d: CubeDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: CubeDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)
        .style("fill", props.barColor)
    boxes
        .transition()
        .duration(500)
        .attr("x", xaxisScale(0))
        .attr("width", function (d: CubeDatum) {
            return xaxisScale(d[countMetric]);
        })
        .attr("y", function (d: CubeDatum) {
            return yaxisScale(d.member);
        })
        .attr("height", yaxisScale.bandwidth() - 1)

    boxes.exit().remove();
}

