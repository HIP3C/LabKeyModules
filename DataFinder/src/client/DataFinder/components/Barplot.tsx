import * as React from 'react';
import { drawBarplot } from './d3/Barplot.d3'
import { CubeDatum } from '../../typings/CubeData'

export interface BarplotProps {
    data: CubeDatum[];
    name: string;
    width: number;
    height: number;
    dataRange: number[];
}


const createAxisData = (data: CubeDatum[]) => {
    const axisData = data.map(cd => cd.member)

    axisData.sort()

    return (axisData)
}

// render the d3 barplot element
export const Barplot: React.FC<BarplotProps> = (props) => {
    React.useEffect(() => {
        if (props.data.length > 0) {
        drawBarplot({
            data: props.data,
            name: props.name,
            width: props.width,
            height: props.height,
            dataRange: props.dataRange,
            labels: createAxisData(props.data)
        });
        }
    });


    return (
        <div className={props.name} >
            <svg id={"barplot-container-" + props.name} ></svg>
        </div>
    );
}
