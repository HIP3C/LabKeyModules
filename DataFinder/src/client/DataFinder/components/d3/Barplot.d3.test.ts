import * as d3 from 'd3'
import { D3Barplot } from './Barplot.d3'
import { CubeDatum } from '../../../typings/CubeData';

describe("D3Barplot", () => {

    // setup
    const config = {
        width: 300,
        height: 300,
        totalHeight: 300,
        labels: ["a", "b", "c", "d"],
        countMetric: "studyCount",
        barColor: "#af88e3"
    }
    const data: CubeDatum[] = [
        {
            level: "Age",
            member: "a",
            studyCount: 10
        },
        {
            level: "Age",
            member: "b",
            studyCount: 50
        },
        {
            level: "Age",
            member: "c",
            studyCount: 12
        },
        {
            level: "Age",
            member: "d",
            studyCount: 42
        }
    ]
    const newData: CubeDatum[] = [
        {
            level: "Age",
            member: "a",
            studyCount: 88
        },
        {
            level: "Age",
            member: "b",
            studyCount: 99
        },
        {
            level: "Age",
            member: "c",
            studyCount: 77
        },
        {
            level: "Age",
            member: "d",
            studyCount: 66
        }
    ]

    document.body.innerHTML =
        `<div id="test-age" >
            <div class="df-barplot-title"><h4>Age</h4></div>
            <div id="barplot-container-Age" class="datafinder-barplot">
                <svg></svg>
            </div>
            <div id="xaxis-Age"}>
                <svg></svg>
            </div>
        </div>`

    const oldXaxisScale = d3
        .scaleLinear()
        .domain([0, 50]).nice()
        .range([0, config.width - 15 - 100]);

    const newXaxisScale = d3
        .scaleLinear()
        .domain([0, 99]).nice()
        .range([0, config.width - 15 - 100]);

    D3Barplot.create("Age", data, config)
    const plot = document.getElementById('test-age')

    test("Create", () => {

        const plotTxt = plot.innerHTML
        expect(plotTxt).toMatchSnapshot()
    })

    test("yaxis", () => {
        const yaxis = document.getElementById("yaxis-labels-short-Age")
        const labels = yaxis.getElementsByTagName("text")
        expect(labels.length).toEqual(4)
        expect(labels[1].innerHTML).toEqual("b")
    })

    test("Update", () => {
        D3Barplot.update("Age", data, config)
        const updatedPlot = document.getElementById('test-age')
        const updatedPlotTxt = updatedPlot.innerHTML
        expect(updatedPlotTxt).toMatchSnapshot()
    })

    test("OldData", () => {
        const barplot = document.getElementById('barplotAge')
        const bars = barplot.getElementsByTagName("rect")
        expect(bars.length).toEqual(4)
        expect(parseFloat(bars[0].getAttribute("width"))).toBeCloseTo(oldXaxisScale(10))
    })

    test("NewData", () => {
        D3Barplot.update("Age", newData, config);
        const barplot = document.getElementById('barplotAge')
        const bars = barplot.getElementsByTagName("rect")
        expect(bars.length).toEqual(4)
        expect(parseFloat(bars[0].getAttribute("width"))).toBeCloseTo(oldXaxisScale(10))

        // wait for transition
        setTimeout(() => {
            const barplot = document.getElementById('barplotAge')
            const bars = barplot.getElementsByTagName("rect")
            expect(bars.length).toEqual(4)
            expect(parseFloat(bars[0].getAttribute("width"))).toBeCloseTo(newXaxisScale(88))
        }, 500)


    })
})

