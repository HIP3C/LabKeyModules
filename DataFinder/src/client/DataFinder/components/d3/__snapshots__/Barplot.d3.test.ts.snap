// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D3Barplot Create 1`] = `
"
            <div class=\\"df-barplot-title\\"><h4>Age</h4></div>
            <div id=\\"barplot-container-Age\\" class=\\"datafinder-barplot\\">
                <svg height=\\"250\\" width=\\"300\\" id=\\"svg-barplot-Age\\"><g id=\\"xaxis-grid-Age\\" transform=\\"translate(100, 255)\\"></g><g transform=\\"translate(100,0)\\" id=\\"bars-Age\\" style=\\"overflow:auto\\"></g><g id=\\"yaxis-labels-short-Age\\" transform=\\"translate(100,0)\\"><text x=\\"-5\\" y=\\"229.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">a</text><text x=\\"-5\\" y=\\"166.66666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">b</text><text x=\\"-5\\" y=\\"104.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">c</text><text x=\\"-5\\" y=\\"41.666666666666664\\" text-anchor=\\"end\\" font-size=\\".8em\\">d</text></g><g id=\\"yaxis-labels-Age\\" transform=\\"translate(100,0)\\"></g></svg>
            </div>
            <div id=\\"xaxis-Age\\" }=\\"\\">
                <svg width=\\"300\\" height=\\"40\\"><g class=\\"x axis\\" transform=\\"translate(100, 0)\\"></g><g transform=\\"translate(100, 0)\\"><text x=\\"92.5\\" y=\\"30\\" class=\\"x-axis-title\\" text-anchor=\\"middle\\" font-size=\\"0.8em\\">Studies</text></g></svg>
            </div>
        "
`;

exports[`D3Barplot Update 1`] = `
"
            <div class=\\"df-barplot-title\\"><h4>Age</h4></div>
            <div id=\\"barplot-container-Age\\" class=\\"datafinder-barplot\\">
                <svg height=\\"250\\" width=\\"300\\" id=\\"svg-barplot-Age\\"><g id=\\"xaxis-grid-Age\\" transform=\\"translate(100, 255)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(0.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"-255\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(74.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"-255\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(148.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"-255\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g></g><g transform=\\"translate(100,0)\\" id=\\"bars-Age\\" style=\\"overflow:auto\\"><rect class=\\"rect\\" x=\\"0\\" width=\\"37\\" y=\\"187.5\\" height=\\"61.5\\" style=\\"fill: #af88e3;\\"></rect><rect class=\\"rect\\" x=\\"0\\" width=\\"185\\" y=\\"125\\" height=\\"61.5\\" style=\\"fill: #af88e3;\\"></rect><rect class=\\"rect\\" x=\\"0\\" width=\\"44.4\\" y=\\"62.5\\" height=\\"61.5\\" style=\\"fill: #af88e3;\\"></rect><rect class=\\"rect\\" x=\\"0\\" width=\\"155.4\\" y=\\"0\\" height=\\"61.5\\" style=\\"fill: #af88e3;\\"></rect></g><g id=\\"yaxis-labels-short-Age\\" transform=\\"translate(100,0)\\"><text x=\\"-5\\" y=\\"229.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">a</text><text x=\\"-5\\" y=\\"166.66666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">b</text><text x=\\"-5\\" y=\\"104.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\">c</text><text x=\\"-5\\" y=\\"41.666666666666664\\" text-anchor=\\"end\\" font-size=\\".8em\\">d</text></g><g id=\\"yaxis-labels-Age\\" transform=\\"translate(100,0)\\"><g class=\\"yaxis-long-label-container\\"><text class=\\"label-number\\" x=\\"200\\" y=\\"229.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\">10</text><rect class=\\"big-rect-cover\\" x=\\"-100\\" y=\\"187.5\\" width=\\"300\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><rect class=\\"little-rect-cover\\" x=\\"-100\\" y=\\"187.5\\" width=\\"100\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><text class=\\"label-text\\" x=\\"-5\\" y=\\"229.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\"><tspan x=\\"-5\\" dy=\\"0\\">a</tspan></text></g><g class=\\"yaxis-long-label-container\\"><text class=\\"label-number\\" x=\\"200\\" y=\\"166.66666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\">50</text><rect class=\\"big-rect-cover\\" x=\\"-100\\" y=\\"125\\" width=\\"300\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><rect class=\\"little-rect-cover\\" x=\\"-100\\" y=\\"125\\" width=\\"100\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><text class=\\"label-text\\" x=\\"-5\\" y=\\"166.66666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\"><tspan x=\\"-5\\" dy=\\"0\\">b</tspan></text></g><g class=\\"yaxis-long-label-container\\"><text class=\\"label-number\\" x=\\"200\\" y=\\"104.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\">12</text><rect class=\\"big-rect-cover\\" x=\\"-100\\" y=\\"62.5\\" width=\\"300\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><rect class=\\"little-rect-cover\\" x=\\"-100\\" y=\\"62.5\\" width=\\"100\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><text class=\\"label-text\\" x=\\"-5\\" y=\\"104.16666666666666\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\"><tspan x=\\"-5\\" dy=\\"0\\">c</tspan></text></g><g class=\\"yaxis-long-label-container\\"><text class=\\"label-number\\" x=\\"200\\" y=\\"41.666666666666664\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\">42</text><rect class=\\"big-rect-cover\\" x=\\"-100\\" y=\\"0\\" width=\\"300\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><rect class=\\"little-rect-cover\\" x=\\"-100\\" y=\\"0\\" width=\\"100\\" height=\\"62.5\\" fill=\\"transparent\\" stroke=\\"transparent\\"></rect><text class=\\"label-text\\" x=\\"-5\\" y=\\"41.666666666666664\\" text-anchor=\\"end\\" font-size=\\".8em\\" fill=\\"transparent\\"><tspan x=\\"-5\\" dy=\\"0\\">d</tspan></text></g></g></svg>
            </div>
            <div id=\\"xaxis-Age\\" }=\\"\\">
                <svg width=\\"300\\" height=\\"40\\"><g class=\\"x axis\\" transform=\\"translate(100, 0)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(0.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"0\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(74.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"0\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g><g class=\\"tick\\" opacity=\\"0.000001\\" transform=\\"translate(148.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"0\\"></line><text fill=\\"currentColor\\" y=\\"3\\" dy=\\"0.71em\\"></text></g></g><g transform=\\"translate(100, 0)\\"><text x=\\"92.5\\" y=\\"30\\" class=\\"x-axis-title\\" text-anchor=\\"middle\\" font-size=\\"0.8em\\">Studies</text></g></svg>
            </div>
        "
`;
